### PR TITLE
Feature/model provider setup

### PR DESCRIPTION
**Pull request by SunnyYasser**
Initial check-in for supporting Azure as a vendor. The changes are as follows

1. The CRUD queries for models have been modified - a new parameter called `provider` is now required to perform crud operations for models
2. A new internal table is added for user defined models. For each operation, first user defined models are searched for the correct (model_name, provider) primary key, and if not found, then default models are searched.
3. The OpenAI header library is split into two - a separate OpenAI client wrapper, and a Session library which uses the libcurl C library.
4. A new AzureModelManager is added to perform the LLM calls for Azure-provided models, and this uses the above Session library
5. Changes are made in the parser for each of the scalar functions to check if provider is specified by the user or not in the input query.

### CODE REVIEW

File: src/core/functions/scalar/llm_filter.cpp [hunk 1]
```diff
@@ -107,13 +109,28 @@ static void LlmFilterScalarFunction2(DataChunk &args, ExpressionState &state, Ve
     Connection con(*state.GetContext().db);
     CoreScalarParsers::LlmFilterScalarParser(args);
 
+    nlohmann::json settings;
+    if (args.ColumnCount() == 4) {
+        settings = CoreScalarParsers::Struct2Json(args.data[3], 1)[0];
+    }
+
+    std::string provider_name = "";
+    bool provider_available = get_provider_name_from_settings (settings, provider_name);
```
Comment (queryproc) [inline src/core/functions/scalar/llm_filter.cpp:118]: you probably want a default back and not a bool. This reads like a weird function and I am uncomfortable mutating inputs within reason. This feels like it should be:
`auto provider_name = get_provider_name(settings, provider_name);`

Comment (queryproc) [reply src/core/functions/scalar/llm_filter.cpp:118]: settings is probably better named as `model_params`.

File: src/core/model_manager/model_manager.cpp [hunk 2]
```diff
@@ -82,25 +85,136 @@ nlohmann::json ModelManager::CallComplete(const std::string &prompt, const std::
...
 }
 
-nlohmann::json ModelManager::CallEmbedding(const std::string &input, const std::string &model) {
-    // List of supported models
-    static const std::unordered_set<std::string> supported_models = {"text-embedding-3-small",
-                                                                     "text-embedding-3-large"};
+nlohmann::json ModelManager::AzureCallComplete(const std::string &prompt, const std::string &model,
+                                                const nlohmann::json &settings, const bool json_response) {
+    // Get API key from the environment variable
+    const char *api_key = AzureModelManager::get_azure_api_key ();
```
Comment (queryproc) [inline src/core/model_manager/model_manager.cpp:91]: I would add the runtime_error exception throwing to the functions. I can forsee any changes to the error requiring multiple places and this leads to repeated code across any azure calls.

File: src/core/utils.cpp [hunk 3]
```diff
@@ -0,0 +1,21 @@
+#include "flockmtl/core/utils.hpp"
+#include <string>
+namespace flockmtl {
+namespace core {
+
+const std::string provider_key = "provider";
+
+static bool helper (const nlohmann::json& json, std::string& provider_name){
```
Comment (queryproc) [inline src/core/utils.cpp:8]: I don't like utils like this. Just put it where it belongs, do:
json.contains(provider_key) ? json.at(provider_key).get<std::string>() : <default_value>

**Review by queryproc — CHANGES_REQUESTED** [review]
small changes requested. Let's discuss if needed!

File: src/core/config/config.cpp [hunk 4]
```diff
@@ -26,9 +35,8 @@ void Config::ConfigSchema(duckdb::Connection &con, std::string &schema_name) {
     }
 }
 
-void Config::ConfigModelTable(duckdb::Connection &con, std::string &schema_name) {
-    const std::string table_name = "FLOCKMTL_MODEL_INTERNAL_TABLE";
-
+void Config::setup_default_models_config (duckdb::Connection &con, std::string &schema_name){
+    const std::string table_name = Config::get_default_models_table_name ();
     // Ensure schema exists
     auto result = con.Query("SELECT table_name FROM information_schema.tables WHERE table_schema = '" + schema_name +
```
[the remark this answers is not in the document]
Comment (SunnyYasser) [reply src/core/config/config.cpp:41]: There is a build conflict with std::format, so we can work on it after the next release?

File: src/core/functions/scalar/llm_embedding.cpp [hunk 5]
```diff
@@ -12,13 +14,29 @@ static void LlmEmbeddingScalarFunction(DataChunk &args, ExpressionState &state,
...
+    std::string provider_name = "";
+    bool provider_available = get_provider_name_from_settings (settings, provider_name);
+
+
     auto model = args.data[1].GetValue(0).ToString();
-    auto query_result =
-        con.Query("SELECT model, max_tokens FROM flockmtl_config.FLOCKMTL_MODEL_INTERNAL_TABLE WHERE model_name = '" +
-                  model + "'");
+
+    auto query_result = provider_available ? con.Query("SELECT model, max_tokens FROM flockmtl_config.FLOCKMTL_MODEL_USER_DEFINED_INTERNAL_TABLE WHERE model_name = '" +
```
[the remark this answers is not in the document]
Comment (SunnyYasser) [reply src/core/functions/scalar/llm_embedding.cpp:28]: yes makes sense, I will change

File: src/core/model_manager/model_manager.cpp (hunk 2 above)
Comment (SunnyYasser) [reply src/core/model_manager/model_manager.cpp:91]: Okay sure will move the error to the correct API

File: src/core/utils.cpp (hunk 3 above)
Comment (SunnyYasser) [reply src/core/utils.cpp:8]: I have added comment in discord regarding this

File: src/core/functions/scalar/llm_embedding.cpp [hunk 6]
```diff
@@ -30,7 +48,8 @@ static void LlmEmbeddingScalarFunction(DataChunk &args, ExpressionState &state,
         for (auto &item : row.items()) {
             concat_input += item.value().get<std::string>() + " ";
         }
-        auto element_embedding = ModelManager::CallEmbedding(concat_input, model_name);
+        auto element_embedding = provider_available ? ModelManager::CallEmbedding(concat_input, model_name, provider_name) :
```
[the remark this answers is not in the document]
Comment (SunnyYasser) [reply src/core/functions/scalar/llm_embedding.cpp:51]: I thought of using "default" as the default provider_name, but then the model selection queries would have had to be modified, and we would also have to define what the default provider is - openai or azure?

I will make changes for point 2

File: src/core/config/config.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (queryproc) [reply src/core/config/config.cpp:41]: Let's open an issue for it.

File: src/core/functions/scalar/llm_embedding.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (queryproc) [reply src/core/functions/scalar/llm_embedding.cpp:51]: What I meant by default is a default that specifies that the provider name is not provided such as null or empty string.

File: src/core/config/config.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (SunnyYasser) [reply src/core/config/config.cpp:41]: done

File: src/core/config/config.cpp [hunk 7]
```diff
@@ -37,17 +45,41 @@ void Config::ConfigModelTable(duckdb::Connection &con, std::string &schema_name)
...
+                  "('default', 'gpt-4o-mini', 'openai', 128000),"
+                  "('gpt-4o-mini', 'gpt-4o-mini', 'openai', 128000),"
+                  "('gpt-4o', 'gpt-4o', 'openai', 128000)");
     }
 }
 
+void Config::setup_user_defined_models_config (duckdb::Connection &con, std::string &schema_name){
+    const std::string table_name = Config::get_user_defined_models_table_name ();
+    // Ensure schema exists
+    auto result = con.Query("SELECT table_name FROM information_schema.tables WHERE table_schema = '" + schema_name +
```
[the remark this answers is not in the document]
Comment (SunnyYasser) [reply src/core/config/config.cpp:63]: done

File: src/core/utils.cpp (hunk 3 above)
Comment (SunnyYasser) [reply src/core/utils.cpp]: done

File: src/core/model_manager/model_manager.cpp (hunk 2 above)
Comment (SunnyYasser) [reply src/core/model_manager/model_manager.cpp]: done

File: src/core/functions/scalar/llm_filter.cpp (hunk 1 above)
Comment (SunnyYasser) [reply src/core/functions/scalar/llm_filter.cpp]: done

**Review by queryproc — APPROVED** [review]

**Pull request merged**